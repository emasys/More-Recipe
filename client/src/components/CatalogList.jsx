import { Link } from 'react-router-dom';
import approx from 'approximate-number';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import React from 'react';

/**
 *
 *
 * @param {number} id
 * @returns {object} toggle classes on mouse in
 */
const onHoverIn = id => {
  document.querySelector(`#recipe-${id}`).classList.remove('crop-text');
  document.querySelector(`#recipe-${id}`).classList.add('full-text');
  document.querySelector(`#img-${id}`).classList.add('d-half');
  document.querySelector(`#tag-${id}`).classList.add('d-none');
};

/**
 *
 *
 * @param {number} id
 * @returns {object} toggle classes on mouse out
 */
const onHoverOut = id => {
  document.querySelector(`#img-${id}`).classList.remove('d-half');
  document.querySelector(`#tag-${id}`).classList.remove('d-none');
  document.querySelector(`#recipe-${id}`).classList.remove('full-text');
  document.querySelector(`#recipe-${id}`).classList.add('crop-text');
};
/**
 *
 *
 * @param {object} props
 * @returns {object} list of recipes
 */
const generateList = props => {
  if (props.catalog) {
    if (props.catalog.length < 1) {
      if (props.isLoading) {
        return (
          <div className="text-center error-message">
            <div>
              <img
                className="img-fluid"
                src="https://res.cloudinary.com/emasys/image/upload/v1516439649/mR_2_jwnuce.png"
                alt="logo"
                height="200"
                width="200"
              />
              <h4 className="p-3 m-2 text-center">...Fetching Recipes</h4>
              <img
                src="https://res.cloudinary.com/emasys/image/upload/v1516647862/Facebook-0.9s-200px_sqqnu9.gif"
                width="100"
                height="100"
                alt="loading..."
              />
            </div>
          </div>
        );
      }
      return (
        <div className="text-center error-message">
          <div>
            <img
              className="img-fluid"
              src="https://res.cloudinary.com/emasys/image/upload/v1516439649/mR_2_jwnuce.png"
              alt="logo"
              height="200"
              width="200"
            />
            <h4 className="p-3 m-2 text-center">...Oops</h4>
            <p className="p-3 m-2 text-center">No recipe here</p>
            <p className="p-3 m-2 text-center">
              <Link to="/" className="btn btn-outline-dark hvr-icon-back">
                ...go back home
              </Link>
            </p>
          </div>
        </div>
      );
    }
    return props.catalog.map(item => (
      <div className="row" key={item.id}>
        <div className="col-lg-12 col-sm-12 mb-20 mt-50 col-md-12">
          <div>
            <Link to={`/recipe/${item.id}`} className="hvr-grow-shadow">
              <div
                className="card"
                data-aos="fade-up"
                data-aos-duration="1000"
                key={item.id}
                onMouseEnter={() => onHoverIn(item.id)}
                onMouseLeave={() => onHoverOut(item.id)}
              >
                {props.auth.authInfo.userId === item.userId &&
                  props.showDeleteBtn && (
                    <button
                      className="btn btn-danger btn-sm delete-btn"
                      onClick={event => props.deleteRecipe(event, item)}
                      data-toggle="modal"
                      data-target="#deleteModal"
                    >
                      Delete Recipe
                    </button>
                  )}
                <div id={`img-${item.id}`}>
                  <img
                    className="card-img-top img-box "
                    src={item.foodImg}
                    alt="recipeImage"
                  />
                </div>

                <div className="card-body p-0 text-center social-icons">
                  <span className="tag bg-danger" id={`tag-${item.id}`}>
                    {item.category}
                  </span>
                  <h4 className="card-title custom-bg bg-dark p-2 m-0">
                    {item.name.length > 25 ?
                      item.name.slice(0, 24).concat('...') :
                      item.name}
                  </h4>
                  <div className="card-body p-5 text-left bg-wheat">
                    <p id={`recipe-${item.id}`} className="crop-text">
                      {item.description}
                    </p>
                  </div>
                  <span>
                    <i className="fa fa-heart-o" aria-hidden="true" />
                    {approx(item.favorite)}
                  </span>
                  <span>
                    <i className="fa fa-thumbs-o-up" aria-hidden="true" />
                    {approx(item.upvote)}
                  </span>
                  <span>
                    <i className="fa fa-thumbs-o-down" aria-hidden="true" />
                    {approx(item.downvote)}
                  </span>
                  <span>
                    <i className="fa fa-eye" aria-hidden="true" />
                    {approx(item.views)}
                  </span>
                  <span>
                    <i className="fa fa-comment-o" aria-hidden="true" />
                    {approx(item.comments)}
                  </span>
                </div>
              </div>
            </Link>
          </div>
        </div>
      </div>
    ));
  }
};

/**
 *
 *
 * @param {object} props
 * @returns {object} container function
 */
const CatalogList = props => (
  <div className="row justify-content-center">{generateList(props)}</div>
);

generateList.propTypes = {
  showDeleteBtn: PropTypes.bool.isRequired,
  catalog: PropTypes.object.isRequired,
  auth: PropTypes.object.isRequired,
  isLoading: PropTypes.bool.isRequired
};

const mapStateToProps = state => ({
  isLoading: state.isLoading,
  auth: state.user
});
export default connect(mapStateToProps, null)(CatalogList);
