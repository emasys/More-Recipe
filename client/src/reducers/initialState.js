export default {
  success: true,
  recipes: [
    {
      id: 1,
      name: 'welcome to more recipe',
      ingredients: ['you', 'me'],
      direction: 'just do it',
      description: 'do it again',
      category: 'Pasta',
      foodImg:
        'some picture url',
      userId: 1,
      upvote: 1,
      downvote: 0,
      reactionUp: [],
      reactionDown: [],
      views: 1,
      comments: 0,
      favorite: 0,
      createdAt: '2017-12-25T12:57:07.971Z',
      updatedAt: '2017-12-25T13:01:32.836Z'
    }
  ]
};
