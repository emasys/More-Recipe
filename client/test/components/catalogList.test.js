import React from 'react';
import { shallow } from 'enzyme';
import CatalogList from '../../src/components/CatalogList';
import initialState from '../../src/reducers/initialState';

test('should render catalog list correctly', () => {
  const wrapper = shallow(<CatalogList catalog={initialState} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render catalog list correctly with empty props', () => {
  const wrapper = shallow(<CatalogList catalog={{ success: true, recipes: [] }} />);
  expect(wrapper).toMatchSnapshot();
});
