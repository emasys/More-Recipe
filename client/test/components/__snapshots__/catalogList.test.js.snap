// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render catalog list correctly 1`] = `
<CatalogList
  catalog={
    Array [
      Object {
        "category": "Breakfast",
        "comments": 5,
        "createdAt": "2017-12-25T11:49:01.973Z",
        "description": "just do it",
        "direction": "just do it",
        "downvote": 0,
        "favorite": 2,
        "foodImg": "https://res.cloudinary.com/emasys/image/upload/v1514202541/hse0xuqzyqlxjqwemigv.jpg",
        "id": 1,
        "ingredients": Array [
          "water",
          "oil",
          "pepper",
          "salt",
        ],
        "name": "How to cook yam",
        "reactionDown": Array [],
        "reactionUp": Array [
          1,
          3,
        ],
        "recipeCount": null,
        "searchIng": null,
        "updatedAt": "2018-01-25T06:05:27.480Z",
        "upvote": 2,
        "userId": 1,
        "views": 3,
      },
    ]
  }
>
  <div
    className="row justify-content-center"
  >
    <div
      className="col-lg-3 col-sm-8 mb-20 col-md-4 animate-catalog"
      data-animate="bounceIn"
      data-delay="0.1s"
      data-duration="1.0s"
      data-offset="100"
      key="0"
    >
      <div>
        <Link
          className=" hvr-grow-shadow"
          replace={false}
          to="/recipe/1"
        />
      </div>
    </div>
  </div>
</CatalogList>
`;

exports[`should render catalog list correctly with empty props 1`] = `
<CatalogList
  catalog={Array []}
>
  <div
    className="row justify-content-center"
  >
    <div
      className="text-center error-message"
    >
      <div
        className="catalog "
      >
        <img
          alt="logo"
          src="https://res.cloudinary.com/emasys/image/upload/v1516439649/mR_2_jwnuce.png"
        />
        <h4
          className="p-3 m-2 text-center"
        >
          ...Oops
        </h4>
        <p
          className="p-3 m-2 text-center"
        >
          No recipe found
        </p>
        <p
          className="p-3 m-2 text-center"
        >
          <Link
            className="btn btn-outline-dark hvr-icon-back"
            replace={false}
            to="/"
          />
        </p>
      </div>
    </div>
  </div>
</CatalogList>
`;
